servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/publisher/PublisherConcatWithCompletableTest.java

```diff
@@ -78,27 +105,70 @@ void nextError() {
         verifySubscriberErrored();
     }
 
-    @Test
-    void sourceCancel() {
+    @ParameterizedTest(name = "{displayName} [{index}] propagateCancel={0} onError={1}")
+    @CsvSource(value = {"false,false", "false,true", "true,false", "true,true"})
+    void sourceCancel(boolean propagateCancel, boolean onError) {
+        setup(propagateCancel);
         subscriber.awaitSubscription().cancel();
         assertThat("Source subscription not cancelled.", subscription.isCancelled(), is(true));
-        assertThat("Next source subscribed on cancellation.", completable.isSubscribed(), is(false));
-        source.onComplete();
-        assertThat("Next source not subscribed.", completable.isSubscribed(), is(true));
-        completable.onSubscribe(cancellable);
-        assertThat("Next cancellable not cancelled.", cancellable.isCancelled(), is(true));
+        assertThat(completable.isSubscribed(), is(propagateCancel));
+        if (onError) {
+            source.onError(DELIBERATE_EXCEPTION);
+        } else {
+            source.onComplete();
+        }
+        if (propagateCancel) {
+            completable.awaitSubscribed();
+            assertThat(cancellable.isCancelled(), is(true));
+            if (onError) {
+                completable.onError(DELIBERATE_EXCEPTION);
+            } else {
+                completable.onComplete();
+            }
+            // Cancel before the publisher completes means we don't propagate terminals.
+            assertThat(subscriber.pollTerminal(10, MILLISECONDS), is(nullValue()));
```

Comment:
Ok, now I see the problem with propagating terminals after `cancel`. Thanks for this test!